### DISCUSSION TITLE
Feature request: C# 7.3 enum generic constraint not being able to compare to default()

### DISCUSSION THREAD

**Discussion by douglasg14b** [dotnet/csharplang#2694] [Q&A] [answered]
I am unable to compare a restricted type parameter to it's default().

Rough Example:

```cs
public class MyClass<TEnum> where TEnum : System.Enum
    {
        public bool SomeMethod(TEnum myEnum)
        {
            return default(TEnum) == myEnum;
        }
}
```

This throws the compiler error: 
> Error CS0019 Operator '==' cannot be applied to operands of type 'TEnum' and 'TEnum'

Do I need to do something else?

**Comment by CyrusNajmabadi** [comment, marked as answer]
> Do I need to do something else?

That code will not work since the size of the enum isn't known at compile time (and the type literally could be `System.Enum` not a subclass), so there's no way to know which `==` to call.  You can replace the above with `default(TEnum).Equals(myEnum)` though.

**Comment by douglasg14b** [comment]
@CyrusNajmabadi 

Thanks for the explanation.

Are there additional generic constraints that could clear that up at compile time? Either existing or proposed/planned?

**Comment by bbarry** [comment]
Does `default(TEnum).Equals(myEnum)` box? And it could throw an exception if you instantiated it as `new MyClass<System.Enum>()`...

I think you want this:

```csharp
using System.Collections.Generic;

public class MyClass<TEnum> where TEnum : struct, System.Enum
    {
        public bool SomeMethod(TEnum myEnum)
        {
            return EqualityComparer<TEnum>.Default.Equals(default, myEnum);
        }
}
```

**Comment by yaakov-h** [comment]
> Does `default(TEnum).Equals(myEnum)` box?

It sure seems like it. [sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDZICYgNQA+AAgMwAERATGQLICeAwhtFADwAqAogHYCuYAPjIB3ABaI4ZLnzBkQFAIwA6HvwCwAKADemsnorlgAeyMYyAZSNg4NODFFHsACmn8yYOqrABKXfp0a+kEUAOxk2HAAZhC8GDAuXt4qAI68EBhQTh6JANx+egC+mgVAA===)

**Comment by CyrusNajmabadi** [comment]
> Are there additional generic constraints that could clear that up at compile time? Either existing or proposed/planned?

It's possible that the 'Shapes' proposal would cover this.
